### PR TITLE
Avoid nan from dzError() in track PV assignment (solves #21780) - 94X version

### PR DESCRIPTION
**Pull request by arizzi**
This PR protects PV assignment from nan replacing the nan with a fixed value (1 meter) for the dZerror

Should solve issue
#21780

Backport of #21816

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc [hunk 1]
```diff
@@ -37,7 +37,9 @@ PrimaryVertexAssignment::chargedHadronVertex( const reco::VertexCollection& vert
     time = 0.;
     timeReso = -1.;
   }
-  
+  float dzError = track->dzError();
+  if(isnan(dzError)) dzError=100; //a nan in dZerror likely means a badly measured track, assume large dz error
```
Comment (slava77) [inline CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc]: please add a LogWarning here

Comment (arizzi) [reply CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc]: there are enough warning already printed out by tracking for that track. 
I think this should be fixed in tracking anyhow and this is just a protection. I will not spend any more time on this PR.

Comment (slava77) [reply CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc]: what are the warnings? 
Looking at http://cms-unified.web.cern.ch/cms-unified/joblogs/vlimant_ACDC0_Run2017F-v1-Tau-17Nov2017_940_171218_115258_6053/76/DataProcessing/70059d5a-e3e3-11e7-b15a-02163e01c3a1-64-3-logArchive/cmsRun1/cmsRun1-stdout.log
I only see "local error not pos-def", but this is rather non-descriptive and also happens frequently in other cases during normal running.

The reason I ask for a warning (or even a LogError) is to have a chance to actually be able to fix this issue in tracking, especially if we can not do it based on this reported crash in this specific release (which may as well be not the only reason bad values can happen).

Comment (arizzi) [reply CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc]: the error being not pos-def is the issue that creates the nan (sqrt of a negative number), that's the thing to monitor and/or fix.
Why it should be up to the PVAssociator to monitor that the final track collection has positive defined error matrix? If you want such protection I suggest you ask TRK POG to insert a check before they push back the tracks in the generalTrack vector.

Comment (slava77) [reply CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc]: I can repeat what I said before: ``` "local error not pos-def", but this is rather non-descriptive and also happens frequently in other cases during normal running.```

Comment (arizzi) [reply CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc]: and I repeat too, if it is non descriptive, ask TRK POG to make it more descriptive, if you want a check more specific on "is the track put in the event having not pos-def error matrix" then put such a check in the track producer, not in a random producer reading tracks...

Comment (slava77) [reply CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc]: fine, but then based on this logic (put a check in the track producer) this PR can not go in.

Comment (arizzi) [reply CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc]: your (O&C) call for this PR not to go in (so you rather keep the crash in production?)
This is a bugfix and I implemented it. The algorithm works perfectly fine and gives the expected results when dzError is not well defined (i.e. nan) so I do not see why this producer should give an Error or Warning.

**Comment by slava77** [comment]
> Should solve issue #21780

Sorry, I'm a bit slow resolving the scope of this bug fix.
I thought that this was a quicker fix yesterday.
IIUC, this applies to the code that runs in the RECO step and the RECO jobs that made the output with the baseline version have completed OK and will probably be not easy to rerun/rereco.
So, by itself, this fix will not solve  #21780 which is an issue in the miniAOD production workflow.

File: CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc (hunk 1 above)
Comment (slava77) [reply CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc]: I have suggested a quick fix and I have asked for a LogWarning to indicate that this after all is not the right place to fix the problem.
I agree that the more appropriate place is to fix the problem in the track producer.
It seems like this fix will not solve the miniAOD crash anyways without rerunning the RECO step upstream.

**Comment by arizzi** [comment]
the config I was given was producing both AOD and MiniAOD so when it crashed also the RECO for those events is lost.

**Comment by slava77** [comment]
On 1/10/18 7:37 AM, arizzi wrote:
> the config I was given was producing both AOD and MiniAOD so when it
> crashed also the RECO for those events is lost.
> 

Ah, good to know.
Thanks.

**Comment by slava77** [comment]
BTW, going back to what I have actually asked for originally, since the discussion has apparently diverged to what was implemented  instead.

In https://github.com/cms-sw/cmssw/issues/21780#issuecomment-356281678
> IIUC, from the messages earlier, the design of PVAssignment is never to return -1.
It seems like it should be independent of the values in the tracks.
Let's have a fix to this.

Can this be done? Do we still have legitimate reasons to return "-1" and the problem with BadRefCore will possibly come back to us due to some other issues in the inputs or eslewhere?

**Comment by arizzi** [comment]
the only way to get again -1 is if the dz() itself is a NAN.. which I think is just not possible.

**Comment by slava77** [comment]
> the only way to get again -1 is if the dz() itself is a NAN.. which I think is just not possible.

can we then make it explicit and never return "-1" in PrimaryVertexAssignment::chargedHadronVertex and return something else (0?) 
or send an exception or assert, since the result is apparently claimed to be impossible?

**Comment by slava77** [comment]
-1

superseded by #21989

**Comment by fabiocos** [comment]
@slava77 @arizzi can we just close this PR, if it is superseded by something already merged?

**Comment by slava77** [comment]
yes, please close

**Pull request closed without merging**